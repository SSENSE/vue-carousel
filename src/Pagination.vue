<template>
  <div v-show="parentContainer.pageCount > 1" class="VueCarousel-pagination">
    <div class="VueCarousel-dot-container">
      <div
        class="VueCarousel-dot"
        v-bind:class="{ 'VueCarousel-dot--active': (index === parentContainer.currentPage) }"
        v-for="(page, index) in parentContainer.pageCount"
        v-on:click="parentContainer.goToPage(index)"
        :style="`
          margin-top: ${parentContainer.paginationPadding * 2}px;
          padding: ${parentContainer.paginationPadding}px;
        `"
      >
        <div
          class="VueCarousel-dot-inner"
          :style="`
            width: ${parentContainer.paginationSize}px;
            height: ${parentContainer.paginationSize}px;
            background: ${(index === parentContainer.currentPage) ? parentContainer.paginationActiveColor : parentContainer.paginationColor};
          `"
        ></div>
      </div>
    </div>
  </div>
</template>

<script>
  export default {
    name: "pagination",
    data() {
      return {
        parentContainer: this.$parent,
      }
    },
  }
</script>

<style scoped>
  .VueCarousel-pagination {
    width: 100%;
    float: left;
    text-align: center;
  }

  .VueCarousel-dot-container {
    display: inline-block;
    margin: 0 auto;
  }

  .VueCarousel-dot {
    float: left;
    cursor: pointer;
  }

  .VueCarousel-dot-inner {
    border-radius: 100%;
  }
</style>
