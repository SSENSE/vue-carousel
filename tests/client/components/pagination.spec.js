import { shallowMount } from '@vue/test-utils';
import Pagination from '../../../src/Pagination';

describe('Pagination Component', () => {
  let carousel;

  beforeEach(() => {
    carousel = {
      pageCount: 1,
      paginationPadding: 1,
      paginationSize: 1,
      paginationActiveColor: 'red',
      paginationColor: 'blue',
      paginationPosition: 'top',
      slideCount: 2,
      scrollPerPage: true,
      currentPerPage: 2,
      currentPage: 0,
      $children: [
        {
          title: 'title1'
        }
      ]
    };
  });

  it('should match the stored snapshot', () => {
    const wrapper = shallowMount(Pagination, { provide: { carousel } });

    expect(wrapper).toMatchSnapshot();
  });

  it('should not contain a button', () => {
    carousel.pageCount = 0;

    const wrapper = shallowMount(Pagination, { provide: { carousel } });

    expect(wrapper.find('button').exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain a button', () => {
    const wrapper = shallowMount(Pagination, { provide: { carousel } });

    expect(wrapper.find('button').exists()).toBe(true);
  });

  it('should set the correct title', () => {
    const wrapper = shallowMount(Pagination, { provide: { carousel } });

    expect(wrapper.vm.getDotTitle(0)).toBe('title1');
  });

  it('should emit paginationclick button event with index 0 when button is clicked', () => {
    const wrapper = shallowMount(Pagination, { provide: { carousel } });

    wrapper.find('button').trigger('click');

    expect(wrapper.emitted().paginationclick[0][0]).toBe(0);
  });

  describe('paginationCount', () => {
    it('should be set to pageCount if scrollPerPage is enabled', () => {
      carousel.scrollPerPage = true;
      carousel.pageCount = 1;

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationCount).toBe(1);
    });

    it('should be set to 0 if the slideCount is undefined and scrollPerPage is disabled', () => {
      carousel.scrollPerPage = false;
      carousel.slideCount = undefined;

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationCount).toBe(0);
    });

    it('should be set to 1 if the slideCount is 1 and scrollPerPage is disabled', () => {
      carousel.scrollPerPage = false;
      carousel.slideCount = 1;

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationCount).toBe(1);
    });
  });

  describe('isCurrentDot', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(Pagination, { provide: { carousel } });
    });

    it('should be false when not the current dot', () => {
      expect(wrapper.vm.isCurrentDot(1)).toBe(false);
    });

    it('should be true when it is the current dot', () => {
      expect(wrapper.vm.isCurrentDot(0)).toBe(true);
    })
  });

  // TODO: What is with this expected behaviour?
  describe('paginationPropertyBasedOnPosition', () => {
    // `bottom`, `bottom-overlay`, `top` and `top-overlay`
    it('should be set to bottom if paginationPosition is set to top', () => {
      carousel.paginationPosition = 'top';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPropertyBasedOnPosition).toBe('bottom');
    });

    it('should be set to bottom if paginationPosition is set to top-overlay', () => {
      carousel.paginationPosition = 'top-overlay';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPropertyBasedOnPosition).toBe('bottom');
    });

    it('should be set to top if paginationPosition is set to bottom', () => {
      carousel.paginationPosition = 'bottom';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPropertyBasedOnPosition).toBe('top');
    });

    it('should be set to top if paginationPosition is set to bottom-overlay', () => {
      carousel.paginationPosition = 'bottom-overlay';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPropertyBasedOnPosition).toBe('top');
    });
  });

  describe('paginationPositionModifierName', () => {
    it('should be the paginationPosition if set to top-overlay', () => {
      carousel.paginationPosition = 'top-overlay';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPositionModifierName).toBe('top-overlay');
    });

    it('should be the paginationPosition if set to bottom-overlay', () => {
      carousel.paginationPosition = 'bottom-overlay';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPositionModifierName).toBe('bottom-overlay');
    });

    it('should be undefined if set to top', () => {
      carousel.paginationPosition = 'top';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPositionModifierName).toBeUndefined();
    });

    it('should be undefined if set to bottom', () => {
      carousel.paginationPosition = 'bottom';

      const wrapper = shallowMount(Pagination, { provide: { carousel } });

      expect(wrapper.vm.paginationPositionModifierName).toBeUndefined();
    });
  });
});
