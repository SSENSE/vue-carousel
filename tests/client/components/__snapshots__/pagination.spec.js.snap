// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component should match the stored snapshot 1`] = `
<div class="VueCarousel-pagination" style="display: none;">
  <div role="tablist" class="VueCarousel-dot-container" style="overflow: hidden; margin: 0px auto; white-space: nowrap;"><button aria-hidden="false" role="tab" title="title1" value="title1" aria-label="title1" aria-selected="true" class="VueCarousel-dot VueCarousel-dot--active" style="margin-bottom: 2px; padding: 1px; width: 1px; height: 1px; background-color: red; transform: translate3d(NaNpx,0,0);"></button></div>
</div>
`;

exports[`Pagination Component should not contain a button 1`] = `
<div class="VueCarousel-pagination" style="display: none;">
  <div role="tablist" class="VueCarousel-dot-container" style="overflow: hidden; margin: 0px auto; white-space: nowrap;"></div>
</div>
`;
